### PR TITLE
Read lines

### PR DESCRIPTION
**Pull request by qwwdfsad**

### CODE REVIEW

File: core/commonMain/src/kotlinx/io/Outputs.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    /** @suppress */
+    override fun close() {
+        // Nothing by default
+    }
+
+    @UseExperimental(ExperimentalStdlibApi::class)
+    private fun ensureCapacity(length: Int) {
+        if (array.size < size + length) {
+            val minCapacity = size + length
+            val powOf2 =
```
Comment (e5l) [inline core/commonMain/src/kotlinx/io/Outputs.kt]: Consider extracting function here

File: core/commonMain/src/kotlinx/io/text/TextInputObsolete.kt [hunk 2]
```diff
@@ -64,34 +65,34 @@ fun Input.readUTF8LineTo(out: Appendable) {
...
-private inline fun Input.decodeUTF8Chars(consumer: (Char) -> Boolean): Int {
+private inline fun Input.decodeUtf8Chars(consumer: (Char) -> Boolean): Int {
     var byteCount = 0
     var value = 0
-    var state = STATE_UTF8
+    var state = STATE_UTF_8
     var count = 0
 
-    while (state != STATE_FINISH) {
+    while (state != STATE_FINISH && !eof()) {
```
Comment (e5l) [inline core/commonMain/src/kotlinx/io/text/TextInputObsolete.kt]: Why we expect `eof` here?

File: core/commonMain/src/kotlinx/io/text/TextInputObsolete.kt [hunk 3]
```diff
@@ -37,25 +38,25 @@ fun Input.readUTF8StringTo(out: Appendable, length: Int): Int {
 /**
  * @throws MalformedInputException
  */
-fun Input.readUTF8Line(): String = buildString {
-    readUTF8LineTo(this)
-}
+//fun Input.readUtf8Line(): String = buildString {
```
Comment (e5l) [inline core/commonMain/src/kotlinx/io/text/TextInputObsolete.kt]: Consider removing this method

File: core/commonMain/src/kotlinx/io/text/TextInput.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+/**
+ * Iterates through each line of the input, calls [action] for each line read
+ * and closes the input when it is exhausted.
+ *
+ * @throws EOFException if the input is already exhausted
+ */
+public inline fun Input.forEachUtf8Line(action: (String) -> Unit) {
+    if (eof()) throw EOFException("Unexpected EOF while reading UTF-8 line")
```
Comment (e5l) [inline core/commonMain/src/kotlinx/io/text/TextInput.kt]: Consider using `Input.checkExhausted()` here.

Comment (e5l) [reply core/commonMain/src/kotlinx/io/text/TextInput.kt]: Throwing on empty input looks strange to me. Should we consider the use cases first?

File: core/commonMain/src/kotlinx/io/Outputs.kt (hunk 1 above)
Comment (qwwdfsad) [reply core/commonMain/src/kotlinx/io/Outputs.kt]: `ensureCapacity` is already a separate method

File: core/commonMain/src/kotlinx/io/text/TextInputObsolete.kt (hunk 2 above)
Comment (qwwdfsad) [reply core/commonMain/src/kotlinx/io/text/TextInputObsolete.kt]: To allow reading strings that are not terminated with a line break.
Frankly speaking, we should also check that eof didn't happen in the middle of the code point, but I decided not to touch the implementation of text decoding at all for now

File: core/commonMain/src/kotlinx/io/text/TextInput.kt (hunk 4 above)
Comment (qwwdfsad) [reply core/commonMain/src/kotlinx/io/text/TextInput.kt]: Good point, `forEachUtf8Line` should be a no-op for empty and exhausted inputs.
My initial idea seems to be flawed

File: core/commonMain/src/kotlinx/io/Outputs.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                if (minCapacity >= 1 shl 30) minCapacity else (1 shl (32 - (minCapacity - 1).countLeadingZeroBits()))
```
Comment (EdwarDDay) [inline core/commonMain/src/kotlinx/io/Outputs.kt]: Can
```
(1 shl (32 - (minCapacity - 1).countLeadingZeroBits()))
```
be replaced by
```
(minCapacity - 1).takeHighestOneBit() shl 1
```
?
This looks easier to understand IMHO.

**Pull request merged**